### PR TITLE
Add parse_index function for parsing NDIndex from string

### PR DESCRIPTION
**Pull request by telamonian**
See #86 for discussion.

Adds a `literal_eval_index` function, built off of a modified version of the `ast.literal_eval` function from the cpy std lib.

I've tested the `literal_eval_index` function in both cpy37 and cpy38, seems to be working pretty well. Still needs some work to integrate this into the `ndindex` package (including adding tests).

### CODE REVIEW

**Comment by telamonian** [comment]
I've now also tested this code in cpy39 (after some teething issues, I managed to use `pyenv` plus some json hackery to get isolated jupyter kernels for each of cpy 37, 38, and 39), and it seems to be working well. Basically, the `ast` authors removed a bunch of classes from `ast` in cpy38 and cpy39 (eg `ast.Num`), but they left in class stubs so my calls to eg `isinstance(x, ast.Num)` don't cause errors in any version.

**Comment by telamonian** [comment]
I've now added unittests for `literal_eval_index`. Locally they're passing with 100% coverage.

There's still a couple of minor flaws in how I'm using hypothesis to generate string representations of ellipses (always`'Ellipsis'`, never `'...'`) and slices (always `'slice(...)'`, never `'x:y'`), but all of those cases are being covered by my `@example`s. Shouldn't be too hard to improve that.

Hypothesis is tricky! And interesting. I've never used it before

File: ndindex/literal_eval_index.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import ast
```
Comment (scopatz) [inline ndindex/literal_eval_index.py:1]: Please add a module level docstring

**Comment by scopatz** [comment]
The code looks reasonable to me.  @asmeurer - will you take a look at this please, whenever you get a chance? Thanks!

File: ndindex/tests/test_literal_eval_index.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+@example('slice(None)')
+@example('slice(None, None)')
+@example('slice(None, None, None)')
+@example('slice(14)')
+@example('slice(12, 14)')
+@example('slice(12, 72, 14)')
+@example('slice(-12, -72, 14)')
+@example('3:15, -5, slice(12, -14), (1,2,3)')
+@example('..., -5, slice(12, -14), (1,2,3)')
+@example('3:15, -5, slice(12, -14), (1,2,3), Ellipsis')
```
Comment (asmeurer) [inline ndindex/tests/test_literal_eval_index.py:49]: These are not valid indices, at least as far as ndindex is concerned. Nested tuples are not allowed.

**Comment by asmeurer** [comment]
Thanks for doing the work on this. This is looking good. Some comments:

- eval_literal_index should pass the result to `ndindex()` so that it returns an ndindex type. The main benefit of this is that you will get type checking on the index that you wouldn't get without doing so. For example, currently nonsense like `literal_eval_index('slice(slice(1, 2))')` is allowed. `literal_eval_index('1.0')` is also allowed. In this way, we don't have to do type checking at the ast level, and can only handle the most basic things. 

- `eval_literal_index` isn't the greatest name. I would prefer something like `parse_index`. 

- This code can go in `ndindex.py`

- The hypothesis tests are fine. You are basically checking that `literal_eval_index(str(index))` works for any ndindex type. I think it might be simpler to do

```py
@given(ndindices)

**Comment by asmeurer** [comment]
Regarding Python 3.9, it is out now, so we need to support it. See https://github.com/Quansight/ndindex/pull/92

File: ndindex/tests/test_literal_eval_index.py (hunk 2 above)
Comment (telamonian) [reply ndindex/tests/test_literal_eval_index.py:49]: I did see

https://github.com/Quansight/ndindex/blob/9e5399fb20f5800474aa2bf081e8323cfeaeabf4/ndindex/tuple.py#L65

Why did you decide to be more restrictive about this syntax than `numpy`?

Comment (asmeurer) [reply ndindex/tests/test_literal_eval_index.py:49]: The reason is that I wanted to avoid confusion from both `Tuple(1, 2)` and `Tuple((1, 2))` being valid and meaning different things. See #17. 

I also had thought that using a tuple to represent an array was deprecated, but now I can't find that.  Certainly there is ambiguity even at the NumPy level. If you expect a tuple to act like a fancy index, then you will be surprised that `a[array([1, 2])]` will give a different thing than `a[(1, 2)]`.

Comment (asmeurer) [reply ndindex/tests/test_literal_eval_index.py:49]: Particularly this comment https://github.com/Quansight/ndindex/issues/17#issuecomment-670772519. Perhaps this should be added to the docs. I do have https://quansight.github.io/ndindex/type-confusion.html#tuple, but that is more about best practices and doesn't really discuss why things are the way they are.

Comment (telamonian) [reply ndindex/tests/test_literal_eval_index.py:49]: Okay, so it will be a little difficult to implement disallowing nested tuples indexes, since there are a bunch of places where we do need to parse `ast.Tuple` regardless. I think the way to do this will be to include some kind of guard in the `ast.Tuple` handling clause of `_convert` that will raise an error if it gets called twice. So far I've got:

```python
class _Guard:
    def __init__(self):
        self.val = False

    def __call__(self):
        if self.val:
            return True
        else:
            self.val = True
            return False
```

and then

```python
    _nested_tuple_guard = _Guard()
    def _convert(node):
        if isinstance(node, ast.Tuple):
            if _nested_tuple_guard():
                raise ValueError(f'tuples inside of tuple indices are not supported: {node!r}')

            return tuple(map(_convert, node.elts))
    ...
```

There's probably a cleaner way to implement the guard

Comment (asmeurer) [reply ndindex/tests/test_literal_eval_index.py]: I think the simpler way is to not worry about nonsense combinations here, and let `ndindex()` handle it at the end.

**Comment by telamonian** [comment]
> Actually I think it needs to look like this
> 
> ```python
> @given(ndindices)
> def test_literal_eval_index(idx):
>     index = ndindex(idx)
>     str_index = str(index)
>     repr_index = repr(index)
>     try:
>         assert literal_eval_index(str_index) == idx
>     except SyntaxError:
>         # str() gives invalid syntax for array indices
>         assume(False)
>     assert literal_eval_index(repr_index) == idx
> ```
> 
> This will test both the str and repr forms, and properly skip the test for the str form of an array (which isn't valid Python).

@asmeurer I tried the approach to writing `test_parse_index` that you suggest above, but it fundamentally won't work as-is. If I understand your intention correctly, you want (at least roughly) the `repr/str` ops and the `parse_index` op to be reciprocal w.r.t. equality, right? This will require some reworking of how `str` and `repr` work in `NDIndex` and/or some of its subclasses.

@asmeurer @scopatz Can you please give this PR another review and let me know if anything still needs to change? Since the version of `parse_index` in this PR works and is passing all tests/coverage checks as of right now, my vote would be to pull this PR in (more-or-less) as-is. We could then ensure that `repr/str` and `parse_index` are reciprocal in a future PR which also makes the necessary rationalizations to the behavior of `repr/str` themselves.

**Comment by telamonian** [comment]
@asmeurer I played around with `repr/str` and `parse_index` a bunch, but everything I tried ran into some kind of trouble. Here's a summary of some of the obstacles:

- Most returns from `repr(ndindex(x))` or `str(ndindex(x))` start with the name of an ndindex class (eg `Tuple`, `IntegerArray`, etc) which `parse_index` does not (and probably should not, ideally it will stick to literals) understand

- There's special handling for `repr(ellipsis())` which turns it into an empty plain tuple (ie `()`)

- For array indices, `str(IntegerArray([0]))` is still valid syntax for `parse_index` if the shape of the array index is 0 or 1

Basically, whenever I fixed one issue I ran into another, so all of the complete fixes I came up so far involve radical changes to `.__repr__`, `.__str__`, and/or `.raw` for some/all `NDIndex` subclasses

**Comment by telamonian** [comment]
I'll close/reopen this PR to trigger the new py39 test

**Comment by telamonian** [comment]
The PR is now passing for py39 as well

File: ndindex/ndindex.py [hunk 3]
```diff
@@ -65,6 +66,91 @@ def __init__(self, f):
     def __get__(self, obj, owner):
         return self.f(owner)
 
+class _Guard:
```
Comment (asmeurer) [inline ndindex/ndindex.py:69]: I don't think this guard stuff is needed. The Tuple constructor should catch this after calling ndindex() at the end (and will give a better error message than what this currently gives).

File: ndindex/ndindex.py [hunk 4]
```diff
@@ -65,6 +66,91 @@ def __init__(self, f):
...
+    """
+    if isinstance(node_or_string, str):
+        node_or_string = ast.parse('dummy[{}]'.format(node_or_string.lstrip(" \t")) , mode='eval')
+    if isinstance(node_or_string, ast.Expression):
+        node_or_string = node_or_string.body
+    if isinstance(node_or_string, ast.Subscript):
+        node_or_string = node_or_string.slice
+
+    def _raise_malformed_node(node):
+        raise ValueError(f'malformed node or string: {node!r}, {ast.dump(node)!r}')
```
Comment (asmeurer) [inline ndindex/ndindex.py:93]: The `node` part here isn't particularly helpful alongside the dump.

Also I don't understand why the literal_eval error message says "node or string". How can the string be malformed? If there is a syntax error, it raises a different exception. I like this better:

```suggestion
        raise ValueError('unsupported node when parsing index: {ast.dump(node)}')
```

File: ndindex/tests/helpers.py [hunk 5]
```diff
@@ -24,12 +24,14 @@
...
 
+positive_ints = integers(1, 10)
 nonnegative_ints = integers(0, 10)
 negative_ints = integers(-10, -1)
 ints = lambda: one_of(negative_ints, nonnegative_ints)
+ints_nonzero = lambda: one_of(negative_ints, positive_ints)
 
 def slices(start=one_of(none(), ints()), stop=one_of(none(), ints()),
-           step=one_of(none(), ints())):
+           step=one_of(none(), ints_nonzero())):
```
Comment (asmeurer) [inline ndindex/tests/helpers.py:34]: This includes zero step because it is used to test the slice constructor. Although it would probably be better to use this for the tests that don't actually test that (see the discussion at https://github.com/Quansight/ndindex/issues/46). It may be better to work on this in a separate PR, though.

File: ndindex/tests/test_ndindex.py [hunk 6]
```diff
@@ -1,17 +1,34 @@
...
+    ints(),
+    slices(),
+)).filter(_doesnt_raise)
+
+ndindexStrs = one_of(
+    ellipses(),
+    ints(),
+    slices(),
+    Tuples,
+).map(lambda x: f'{x}')
```
Comment (asmeurer) [inline ndindex/tests/test_ndindex.py:26]: Put strategies in helpers.py

File: ndindex/tests/test_ndindex.py [hunk 7]
```diff
@@ -77,6 +94,63 @@ def test_ndindex_invalid():
...
+@example('slice(None, None, None)')
+@example('slice(14)')
+@example('slice(12, 14)')
+@example('slice(12, 72, 14)')
+@example('slice(-12, -72, 14)')
+@example('3:15, -5, slice(-12, -72, 14), Ellipsis')
+@example('..., 3:15, -5, slice(-12, -72, 14)')
+@given(ndindexStrs)
+def test_parse_index_hypothesis(ixStr):
+    assert eval(f'_dummy[{ixStr}]') == parse_index(ixStr)
```
Comment (asmeurer) [inline ndindex/tests/test_ndindex.py:122]: You can check here if eval gives an error than so should parse_index.

File: ndindex/tests/test_ndindex.py (hunk 6 above)
Comment (asmeurer) [reply ndindex/tests/test_ndindex.py:26]: You can include arrays here too, although you'll have to use the repr form in that case.

**Comment by asmeurer** [comment]
One idea to get better test coverage would be to `__str__` for ndindex types to give the raw string form, like

```py
>>> str(Slice(1, 2))
'1:2'
>>> str(Tuple(1, 2))
'(1, 2)'
```

and then test both str and repr of ndindex(idx) (and ndindex(idx).raw) (str for arrays will still have to be skipped because it uses non-valid syntax). 

It may also be a good idea to support the ndindex classes as valid names, so `parse_index('Slice(1, 2)')` works.

File: ndindex/ndindex.py [hunk 8]
```diff
@@ -65,6 +66,91 @@ def __init__(self, f):
...
+    def __call__(self):
+        if self.val:
+            return True
+        else:
+            self.val = True
+            return False
+
+def parse_index(node_or_string):
+    """
+    "Safely" (needs validation) evaluate an expression node or a string containing
```
Comment (asmeurer) [inline ndindex/ndindex.py:82]: What does "needs validation" here mean?

File: ndindex/ndindex.py [hunk 9]
```diff
@@ -65,6 +66,91 @@ def __init__(self, f):
...
+    a (limited subset) of valid numpy index or slice expressions.
```
Comment (asmeurer) [inline ndindex/ndindex.py:83]: The docstring should probably clarify what is and isn't allowed. I guess we aren't allowing any kind of expressions that create numbers. 

Actually, we may at one point want to support just anything that parses, and make the current behavior under a `safe=True` flag, which I guess should be the default. But we can worry about that later. 

Also this docstring needs examples. and should be added to Sphinx.

File: ndindex/ndindex.py (hunk 8 above)
Comment (asmeurer) [reply ndindex/ndindex.py:82]: I guess it depends on what we mean by "safely". `parse_index('('*1000 + ')'*1000)` raises MemoryError (but so does `ast.literal_eval('('*1000 + ')'*1000)`). It shouldn't be possible to execute arbitrary code, though, as far as I can tell.

Comment (telamonian) [reply ndindex/ndindex.py:82]: doesn't really mean anything, mostly just a reflection of my paranoia about building a) any parser, and b) something meant to take raw user input. My goal has been to be something that's at least as safe as the existing `literal_eval`. At this point I've examined the code and the surrounding issues enough that I'm reasonably confident this has actually been archived, so I'll edit out the wishy-washy language

File: ndindex/ndindex.py (hunk 4 above)
Comment (telamonian) [reply ndindex/ndindex.py:93]: I think your language is clearer in any case. I'll pull this suggestion in

File: ndindex/tests/helpers.py (hunk 5 above)
Comment (telamonian) [reply ndindex/tests/helpers.py:34]: I added this since I kept getting `ValueErrors` due to the related check in the `Slice` constructor.

So what I don't get is how this wasn't causing errors previously for tests using the `@given(ndindices)` strategy. Wouldn't that just mean that (due to one of the many weird quirks of hypothesis) `@given(ndindices)` just happened to never attempt to generate `ndindex(slice(x, y, 0))`?

Comment (asmeurer) [reply ndindex/tests/helpers.py:34]: ndindices uses `filter(_doesnt_raise)` which make it only generate valid indices. Actually I suspect this means several tests
